### PR TITLE
Battery calibration

### PR DESCRIPTION
**Pull request by KyrosWeb**
Allow battery tool to discharge your battery until it reaches 15%. When the 15% is reached, it will fully recharge the battery and keep it at 100% for 1 hour.Finally the battery will be discharged until it reaches 80% and will be kept on this percentage.
From what I understand it would be advisable to calibrate the battery about every 3 months of continuous use at 80%

### CODE REVIEW

**Comment by LorenzoRogai** [comment]
Looking good for me 👍

**Comment by Manoplay** [comment]
Great job, seems good! 👍🏻

File: battery.sh [hunk 1]
```diff
@@ -568,3 +571,36 @@ if [[ "$action" == "logs" ]]; then
...
+if [[ "$action" == "calibrate" ]]; then
+	log "Starting calibration"
+
+	# Stop the maintaining
+	battery maintain stop
+
+	# Discharge battery to 15%
+	battery discharge 15
+
+	while true; do
```
Comment (actuallymentor) [inline battery.sh:585]: For this, please use the built in `battery charge` command. When `battery charge 100` is done, the command exits and lets your script continue

Comment (KyrosWeb) [reply battery.sh:585]: If I use "battery charge 100" won't it start battery discharging after the execution? I need to wait 1 hour before discharging process.

**Comment by KyrosWeb** [comment]
I've done the calibration checking, i wait for others.
Thank you

**Comment by KyrosWeb** [comment]
I've done the calibration checking, i wait for others.
Thank you @actuallymentor

**Comment by KyrosWeb** [comment]
What do you think if during the calibration the magsafe led will blinking in orange? @actuallymentor

File: battery.sh [hunk 2]
```diff
@@ -568,3 +571,36 @@ if [[ "$action" == "logs" ]]; then
...
+			sleep 300
+			continue
+		fi
+	done
+
+	# Wait before discharging to target level
+	log  "reached 100%, maintaining for 1 hour"
+	sleep 3600
+
+	# Discharge battery to 80%
```
[the remark this answers is not in the document]
Comment (KyrosWeb) [reply battery.sh]: Yes but the battery need first to discharge , right?

**Comment by actuallymentor** [comment]
> What do you think if during the calibration the magsafe led will blinking in orange? @actuallymentor 

I think that's a very cool idea!

Do make sure to:

1. Tell the user when they run the command
2. Add the orange blink to the visudo list
3. Add it to the visudo check (see visudo command usage of Sudo -n)

**Comment by KyrosWeb** [comment]
> > What do you think if during the calibration the magsafe led will blinking in orange? @actuallymentor
> 
> I think that's a very cool idea!
> 
> Do make sure to:
> 
> 1. Tell the user when they run the command
> 2. Add the orange blink to the visudo list
> 3. Add it to the visudo check (see visudo command usage of Sudo -n)

I was checking about it but the led is continuously overwritten by other commands. Maybe something for the future!

**Comment by KyrosWeb** [comment]
When it will be integrated?
@actuallymentor

**Comment by odpay** [comment]
looks good to me

**Comment by hexclann** [comment]
Tested on Macbook Air M2, looks good!

File: battery.sh [hunk 3]
```diff
@@ -65,6 +66,9 @@ Usage:
     manually set the adapter to (not) charge even when plugged in
     eg: battery adapter off
 
+  battery calibrate 
+    calibrate the battery by discharding it to 15%, then recharging it to 100%, and keeping it there for 1 hour
```
Comment (actuallymentor) [inline battery.sh:70]: Minor misspelling of `discharding`, should be `discharging`

File: battery.sh [hunk 4]
```diff
@@ -406,6 +410,12 @@ fi
 
 # Maintain at level
 if [[ "$action" == "maintain_synchronous" ]]; then
+	
+	# Checking if the calibration process is running
+	if test -f "$calibrate_pidfile"; then
+		pid=$(cat "$calibrate_pidfile" 2>/dev/null)
+		log "🚨 Calibration is running, please run 'battery calibrate stop' or wait for it to finish"
```
Comment (actuallymentor) [inline battery.sh:417]: This line should probably not only log an error but exit the command so the processes do not conflict.

File: battery.sh [hunk 5]
```diff
@@ -477,6 +487,11 @@ if [[ "$action" == "maintain" ]]; then
 		kill $pid &>/dev/null
 	fi
 
+	if test -f "$calibrate_pidfile"; then
+		pid=$(cat "$calibrate_pidfile" 2>/dev/null)
+		log "🚨 Calibration is running, please run 'battery calibrate stop' or wait for it to finish"
```
Comment (actuallymentor) [inline battery.sh:492]: Same as above, the logic should exit here so clashes do not just cause error logs. During a calibration there should either:

1. Be a blocking of maintenance commands
2. The maintenance commands should stop the calibration and give a verbose output of "hey I stopped the calibration"

File: battery.sh [hunk 6]
```diff
@@ -520,6 +535,64 @@ if [[ "$action" == "maintain" ]]; then
...
+	# Discharge battery to 15%
+	battery discharge 15
+
+	while true; do
+		log "checking if at 100%"
+		# Check if battery level has reached 100%
+		if battery status | head -n 1 | grep -q "Battery at 100%"; then
+			break
+			else
+			sleep 300
```
Comment (actuallymentor) [inline battery.sh:554]: Minor style comment: indent else lines

File: battery.sh [hunk 7]
```diff
@@ -520,6 +535,64 @@ if [[ "$action" == "maintain" ]]; then
...
+			continue
+		fi
+	done
+
+	# Wait before discharging to target level
+	log  "reached 100%, maintaining for 1 hour"
+	sleep 3600
+
+	# Discharge battery to 80%
+	battery discharge 80
```
Comment (actuallymentor) [inline battery.sh:564]: Does this make more sense than running `recover`? Is discharging to 80 part of the calibration or is it based on the assumption that we are aiming to return to a baseline of 80?

**Review by actuallymentor — CHANGES_REQUESTED** [review]
Thank you ver much for this PR @KyrosWeb!

My apologies for only getting to it so late.

I've requested some clarifications/edits, but they are quite minor. Let me know what you think and add edits to the extent that you think they make sense.

I have some bandwidth this week so will respond quicker than I have over the past months :)

File: battery.sh (hunk 5 above)
Comment (KyrosWeb) [reply battery.sh:492]: Also here? Should stop the calibration process?

File: battery.sh (hunk 4 above)
Comment (KyrosWeb) [reply battery.sh:417]: You say , it should stop the battery calibration process?

File: battery.sh (hunk 7 above)
Comment (KyrosWeb) [reply battery.sh:564]: No, the discharging is a part of the calibration.

**Comment by seamusdemora** [comment]
Apologies for "butting in" at this late date, but...

Just out of curiosity, did the authors consult this article: [Testing and Calibrating Smart Batteries](https://batteryuniversity.com/article/bu-605-calibrating-smart-batteries-with-impedance-tracking) in the development of this PR? I ask because I've not seen any mention of "Impedance Tracking"... I wonder if Apple actually uses this technology? It's been available in hardware for a while, but no idea how - or if - Apple has implemented it. 

And please don't misunderstand; I think the PR is a sound idea, but I couldn't help but wonder if Apple might have implemented impedance tracking - and if anyone has given thought as to how to learn if they have - and how to take advantage of it **if** they have.

**Comment by KyrosWeb** [comment]
@seamusdemora Hi , Yes i read it all and i think the process of discharging and recharging the battery when it stay lots of time stuck in a percentage is important to preserve the battery life and in order to avoid these situations : https://github.com/actuallymentor/battery/issues/191
This kind of method is also used by other battery maintaining app and I think mine is the faster and easier way to implement it

**Comment by seamusdemora** [comment]
@KyrosWeb :  
OK - thank you...  I've read through all of the above & think I understand the approach. Is this included the ver 1.2.1 Release?? 

P.S. I liked the approach except for the "blinking light"   :)  IMHO that might be distracting?

**Comment by KyrosWeb** [comment]
@seamusdemora Hi , it will be included soon. I haven’t insert anymore the blinking light  for your happiness 😂

**Comment by seamusdemora** [comment]
@KyrosWeb   That is good to hear. I'm currently evaluating `bclm`, and so I will wait for the next release of `battery` before installing. 

P.S.  Yes, you have made me very happy  :)

**Comment by djbob2000** [comment]
I'm looking forward to the release of this version with calibration:)

**Comment by KyrosWeb** [comment]
> I'm looking forward to the release of this version with calibration:)

We need to wait the repo owner.

**Comment by odpay** [comment]
can we get a merge pls

**Comment by KyrosWeb** [comment]
@actuallymentor Hey, no news for PR?

**Comment by actuallymentor** [comment]
Apologies on the delay, merged!

Thank you for your work @KyrosWeb!

Please keep an eye on the issues tab the coming weeks to make sure to catch any issues :)

**Comment by seamusdemora** [comment]
It's only been 4 ½  months  - there's no rush; you need to take it easy!

**Pull request merged**